agent/api/task_unix.go

```diff
@@ -15,16 +15,127 @@
+	// With a 100ms CPU period, we can express 0.01 vCPU to 10 vCPUs
+	maxTaskVCPULimit = 10
+	// Reference: http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html
+	minimumCPUShare = 2
 )
 
-func (task *Task) adjustForPlatform() {}
+func (task *Task) adjustForPlatform(cfg *config.Config) {
+	task.memoryCPULimitsEnabledLock.Lock()
+	defer task.memoryCPULimitsEnabledLock.Unlock()
+	task.memoryCPULimitsEnabled = cfg.TaskCPUMemLimit
+}
 
 func getCanonicalPath(path string) string { return path }
 
-// CgroupEnabled checks whether cgroups need be enabled at the task level
-func (task *Task) CgroupEnabled() bool {
-	//stub for now
-	return false
+// BuildCgroupRoot helps build the task cgroup prefix
+// Example: /ecs/task-id
+func (task *Task) BuildCgroupRoot() (string, error) {
+	taskID, err := task.GetID()
+	if err != nil {
+		return "", errors.Wrapf(err, "task build cgroup root: unable to get task-id from taskARN: %s", task.Arn)
+	}
+
+	return filepath.Join(config.DefaultTaskCgroupPrefix, taskID), nil
+}
+
+// BuildLinuxResourceSpec returns a linuxResources object for the task cgroup
+func (task *Task) BuildLinuxResourceSpec() (specs.LinuxResources, error) {
+	linuxResourceSpec := specs.LinuxResources{}
+
+	// If task level CPU limits are requested, set CPU quota + CPU period
+	if !utils.ZeroOrNil(task.VCPULimit) {
+		if task.VCPULimit > maxTaskVCPULimit || task.VCPULimit < 0 {
+			return specs.LinuxResources{},
+				errors.Errorf("task resource spec builder: unsupported CPU limits, requested=%f, max-supported=%d",
+					task.VCPULimit, maxTaskVCPULimit)
+		}
+		taskCPUPeriod := uint64(defaultCPUPeriod / time.Microsecond)
+		taskCPUQuota := int64(task.VCPULimit * float64(taskCPUPeriod))
+
+		// TODO: DefaultCPUPeriod only permits 10VCPUs.
+		// Adaptive calculation of CPUPeriod required for further support
+		linuxResourceSpec.CPU = &specs.LinuxCPU{
+			Quota:  &taskCPUQuota,
+			Period: &taskCPUPeriod,
+		}
+	} else {
+		// If task level CPU limits are missing,
+		// aggregate container CPU shares when present
+		var taskCPUShares uint64
+		for _, container := range task.Containers {
+			if !utils.ZeroOrNil(container.CPU) {
+				taskCPUShares += uint64(container.CPU)
+			}
+		}
+
+		// If there are are no CPU limits at task or container level,
+		// default task CPU shares
+		if utils.ZeroOrNil(taskCPUShares) {
+			// Set default CPU shares
+			taskCPUShares = minimumCPUShare
+		}
+
+		linuxResourceSpec.CPU = &specs.LinuxCPU{
+			Shares: &taskCPUShares,
+		}
+	}
+
+	// If task memory limit is not present, cgroup parent memory is not set
+	// If task memory limit is set, ensure that no container
+	// of this task has a greater request
+	if !utils.ZeroOrNil(task.MemoryLimit) {
+		taskMemoryLimit := int64(task.MemoryLimit)
+		for _, container := range task.Containers {
+			containerMemoryLimit := int64(container.Memory)
+			if !utils.ZeroOrNil(containerMemoryLimit) && containerMemoryLimit > taskMemoryLimit {
```

Comment:
What is the `!utils.ZeroOrNil(containerMemoryLimit)` check giving us here that the `>` is not already doing?